contributors/design-proposals/daemonset-update.md

```diff
@@ -0,0 +1,268 @@
+	// number is calculated from percentage by rounding up.
+	// This must be greater than 0.
+	// Default value is 1.
+	// Example: when this is set to 30%, 30% of the currently running DaemonSet
+	// pods can be stopped for an update at any given time. The update starts
+	// by stopping at most 30% of the currently running DaemonSet pods and then
+	// brings up new DaemonSet pods in their place. Once the new pods are ready,
+	// it then proceeds onto other DaemonSet pods, thus ensuring that at least
+	// 70% of original number of DaemonSet pods are available at all times
+	// during the update.
+	MaxUnavailable intstr.IntOrString
+}
+
+// DaemonSetSpec is the specification of a daemon set.
+type DaemonSetSpec struct {
+	// Note: Existing fields, including Selector and Template are ommitted in
+	// this proposal.  
+
+	// Update strategy to replace existing DaemonSet pods with new pods.
+	UpdateStrategy DaemonSetUpdateStrategy `json:"updateStrategy,omitempty"`
+
+	// Minimum number of seconds for which a newly created DaemonSet pod should
+	// be ready without any of its container crashing, for it to be considered
+	// available. Defaults to 0 (pod will be considered available as soon as it
+	// is ready).
+	MinReadySeconds int32 `json:"minReadySeconds,omitempty"`
+
+	// The number of old PodTemplates to retain to allow rollback.
+	// This is a pointer to distinguish between explicit zero and not specified.
+	// Defaults to 3. 
+	RevisionHistoryLimit *int32 `json:"revisionHistoryLimit,omitempty"`
+}
+
+const (
+	// DefaultDaemonSetUniqueLabelKey is the default key of the labels that is added
+	// to daemon set pods to distinguish between old and new pod templates during
+	// DaemonSet update.
+	DefaultDaemonSetUniqueLabelKey string = "pod-template-hash"
+)
+
+// DaemonSetStatus represents the current status of a daemon set.
+type DaemonSetStatus struct {
+	// Note: Existing fields, including CurrentNumberScheduled, NumberMissscheduled,
+	// DesiredNumberScheduled, NumberReady, and ObservedGeneration are ommitted in
+	// this proposal.
+
+	// UpdatedNumberScheduled is the total number of nodes that are running updated
+	// daemon pod
+	UpdatedNumberScheduled int32 `json:"updatedNumberScheduled"`
+}
+```
+
+### Controller 
+
+#### DaemonSet Controller 
+
+The DaemonSet Controller will make DaemonSet updates happen. It will watch
+DaemonSets in etcd. 
+
+For each pending DaemonSet updates, it will:
+
+1. Find all pods whose labels are matched by DaemonSet `.spec.selector`. 
+   - If `OwnerReference` is implemented for DaemonSets, filter out pods that
+     aren't controlled by this DaemonSet too
+1. Find all nodes that should run these pods created by this DaemonSet.
+1. Find existing PodTemplates whose labels are matched by DaemonSet
+   `.spec.selector`
+   - Sort those PodTemplates by creation timestamp and only retain at most
+     `.spec.revisionHistoryLimit` latest PodTemplates (remove the rest)
```

Comment:
**Discussion** - In Deployment controller, we only kill old ReplicaSets (history) whose replicas=0 during clean up. Should we do the same for PodTemplates (only kill history that no pods are referencing)? IMO we don't need to, since unlike Deployment's case, killing PodTemplates won't cause pods with the same template to be killed.